### PR TITLE
RO refactor

### PR DESCRIPTION
**Pull request by bknueven**
## Fixes/Addresses:

(replace this with the issue # fixed or addressed, if no issue exists then a brief statement of what this PR does)

## Changes proposed in this PR:
- [x] Finish unifying 0DRO & 1DRO models to the extent possible
- [x] Update affected flowsheets
- [x] Update 0DRO Jupyter notebook
- [x] Update Documentation, including out-of-date information in "how to scale a model"
- [x] Sneaks in a minor fix to #339 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bknueven** [comment]
@TimBartholomew and @adam-a-a, this is ready for a first-pass review. The tests are going to fail because I haven't changed anything outside the `unit_model` directory. There's also still some minor differences between 0D and 1D RO which could probably be resolved, but I'm not certain how to do it.

@andrewlee94 while this allows us to express the constraints common to these models, I think makes the individual models harder to understand, especially given how we need to make the 0D model "quack" like the 1D model. I would appreciate any comments on how I've implemented the bridging between the two.

File: watertap/unit_models/reverse_osmosis_0D.py [hunk 1]
```diff
@@ -205,46 +86,112 @@ def build(self):
...
-            "property_package": self.config.property_package,
-            "property_package_args": self.config.property_package_args})
-
-        self.permeate_side.add_state_blocks(
-            has_phase_equilibrium=False)
+        # for quacking like 1D model
+        add_object_reference(self.feed_side,'properties',
+                {**{(t,0.) : self.feed_side.properties_in[t] for t in self.flowsheet().config.time},
+                 **{(t,1.) : self.feed_side.properties_out[t] for t in self.flowsheet().config.time}}
+                )
```
Comment (TimBartholomew) [inline watertap/unit_models/reverse_osmosis_0D.py:93]: Nice. I was wondering how we could centralize things across the 0D and 1D stateblocks.

File: watertap/unit_models/reverse_osmosis_0D.py [hunk 2]
```diff
@@ -205,46 +86,112 @@ def build(self):
...
+        solute_set = self.config.property_package.solute_set
+
+        self.rejection_phase_comp = Var(
+            self.flowsheet().config.time,
+            self.config.property_package.phase_list,
+            solute_set,
+            initialize=0.9,
+            bounds=(1e-2, 1 - 1e-6),
+            units=pyunits.dimensionless,
+            doc='Observed solute rejection')
```
Comment (TimBartholomew) [inline watertap/unit_models/reverse_osmosis_0D.py]: Why can't this be moved to the base class? Did the 1DRO model not have rejection?

File: watertap/unit_models/reverse_osmosis_0D.py [hunk 3]
```diff
@@ -259,198 +206,67 @@ def mass_transfer_phase_comp_initialize(b, t, p, j):
...
 
+        # constraints for additional variables (i.e. variables not used in other constraints)
+
+        # Not in 1DRO
+        @self.Constraint(self.flowsheet().config.time,
+                         solute_set)
+        def eq_rejection_phase_comp(b, t, j):
+            return (b.rejection_phase_comp[t, 'Liq', j] ==
+                    1 - (b.mixed_permeate[t].conc_mass_phase_comp['Liq', j] /
+                         b.feed_side.properties_in[t].conc_mass_phase_comp['Liq', j]))
```
Comment (TimBartholomew) [inline watertap/unit_models/reverse_osmosis_0D.py]: Ah ok this answers my previous question. The rejection variable not being in the 1DRO is an oversight, it should also have it.

**Review by TimBartholomew — COMMENTED** [review]
I think it looks good. Just a few comments.

**Comment by TimBartholomew** [comment]
@bknueven which things are "There's also still some minor differences between 0D and 1D RO which could probably be resolved, but I'm not certain how to do it."? It might be best for us to just have a quick meeting to chat about them

**Comment by adam-a-a** [comment]
> @bknueven which things are "There's also still some minor differences between 0D and 1D RO which could probably be resolved, but I'm not certain how to do it."? It might be best for us to just have a quick meeting to chat about them

I can also pitch in on things related to 1DRO (e.g., rejection can and should be included). There may be reasons why I did things differently for 1DRO but seeing the reformulation, I'm not sure about what else @bknueven was referring to regarding the minor differences that could probably be resolved. That said, I still need to review the PR in full.

File: watertap/unit_models/reverse_osmosis_1D.py [hunk 4]
```diff
@@ -202,46 +208,15 @@ def build(self):
...
                 self.config.momentum_balance_type != MomentumBalanceType.none):
-            add_object_reference(self, 'deltaP', feed_side.deltaP)
+            add_object_reference(self, 'dP_dx', feed_side.deltaP)
 
         self._make_performance()
 
         self._add_expressions()
 
+        # TODO: why is this here?
         self._get_performance_contents()
```
Comment (adam-a-a) [inline watertap/unit_models/reverse_osmosis_1D.py]: I don't think it needs to be there. This was one of my earliest PRs so it was probably an oversight.

File: watertap/unit_models/reverse_osmosis_1D.py [hunk 5]
```diff
@@ -463,223 +307,49 @@ def eq_flux_mass(b, t, x, p, j):
...
-                        * bulk.dens_mass_phase['Liq'] * b.velocity[t, x]**2)
         ## ==========================================================================
         # Feed-side isothermal conditions
-
+        # NOTE: this could go on the feed_side block, but that seems to hurt initialization
+        #       in the tests for this unit
         @self.Constraint(self.flowsheet().config.time,
-                         self.feed_side.length_domain,
+                         self.difference_elements,
                          doc="Isothermal assumption for permeate")
```
Comment (adam-a-a) [inline watertap/unit_models/reverse_osmosis_1D.py]:
```suggestion
                         doc="Isothermal assumption for feed channel")
```

File: watertap/unit_models/reverse_osmosis_1D.py [hunk 6]
```diff
@@ -463,223 +307,49 @@ def eq_flux_mass(b, t, x, p, j):
...
-                             doc="pressure change per unit length due to friction")
-            def eq_dP_dx(b, t, x):
-                bulk = b.feed_side.properties[t, x]
-                return (b.deltaP[t, x] * b.dh ==
-                        -0.5 * b.friction_factor_darcy[t, x]
-                        * bulk.dens_mass_phase['Liq'] * b.velocity[t, x]**2)
         ## ==========================================================================
         # Feed-side isothermal conditions
-
+        # NOTE: this could go on the feed_side block, but that seems to hurt initialization
```
Comment (adam-a-a) [inline watertap/unit_models/reverse_osmosis_1D.py:327]: Good to know although I'm not sure why initialization would be adversely affected.

File: watertap/unit_models/reverse_osmosis_1D.py (hunk 4 above)
Comment (bknueven) [reply watertap/unit_models/reverse_osmosis_1D.py]: Removed.

**Comment by bknueven** [comment]
Waiting on #336 for analysis flowsheet tests.

**Comment by bknueven** [comment]
@lbianchi-lbl could you check the changes I made in f90d00c?

File: watertap/examples/flowsheets/lsrro/tests/test_lsrro.py [hunk 7]
```diff
@@ -193,32 +193,37 @@ def test_set_operating_conditions(self, model):
...
         assert captured.out == self.display_system
 
+    @pytest.mark.requires_idaes_solver
     @pytest.mark.component
     def test_display_design(self, model, capsys):
         display_design(model)
         captured = capsys.readouterr()
         assert captured.out == self.display_design
 
+    @pytest.mark.requires_idaes_solver
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/lsrro/tests/test_lsrro.py]: Does this mean we shouldn't use ipopt-watertap when working with LSRRO?

File: docs/technical_reference/unit_models/reverse_osmosis_0D.rst [hunk 8]
```diff
@@ -61,7 +61,7 @@ Variables
    "Solvent permeability coefficient", ":math:`A`", "A_comp", "[t, j]", ":math:`\text{m/Pa/s}`"
    "Solute permeability coefficient", ":math:`B`", "B_comp", "[t, j]", ":math:`\text{m/s}`"
    "Mass density of solvent", ":math:`\rho_{solvent}`", "dens_solvent", "[p]", ":math:`\text{kg/}\text{m}^3`"
-   "Mass flux across membrane", ":math:`J`", "flux_mass_io_phase_comp", "[t, x, p, j]", ":math:`\text{kg/s}\text{/m}^2`"
+   "Mass flux across membrane", ":math:`J`", "flux_mass_phase_comp", "[t, x, p, j]", ":math:`\text{kg/s}\text{/m}^2`"
```
Comment (adam-a-a) [inline docs/technical_reference/unit_models/reverse_osmosis_0D.rst:64]: Does the reverse_osmosis_1D.rst need any updating as well? For example, changing `deltaP` to `dP_dx`?

Comment (bknueven) [reply docs/technical_reference/unit_models/reverse_osmosis_0D.rst:64]: Probably -- it seems I need to take another pass at the documentation. Difficult for me to remember where I left off after a week!

File: watertap/examples/flowsheets/lsrro/tests/test_lsrro.py (hunk 7 above)
Comment (bknueven) [reply watertap/examples/flowsheets/lsrro/tests/test_lsrro.py]: No, the difference is insignificant, but doesn't seem worth resolving. This also only affects those without an IDAES provided Ipopt. Note that `ipopt-watertap` is just a wrapper around *whatever* Ipopt the user has.

On my mac without these xfails:
```
========================================================= short test summary info =========================================================
FAILED tests/test_lsrro.py::TestLSRRO_2Stage::test_initialize - assert 0.2985162459059866 == 0.298029 ± 3.0e-06
FAILED tests/test_lsrro.py::TestLSRRO_3Stage::test_initialize - assert 0.35344241623557743 == 0.351684 ± 3.5e-06
================================================= 2 failed, 25 passed in 60.02s (0:01:00) =================================================
```

But, reconsidering it, I think the better answer is just relaxing the relative tolerance, especially given this is just initialization.

File: docs/technical_reference/unit_models/reverse_osmosis_0D.rst (hunk 8 above)
Comment (bknueven) [reply docs/technical_reference/unit_models/reverse_osmosis_0D.rst:64]: Actually, I checked -- this PR makes the 1DRO documentation "more correct", and doesn't need updates.
https://github.com/watertap-org/watertap/blob/main/docs/technical_reference/unit_models/reverse_osmosis_1D.rst#variables

**Review by TimBartholomew — APPROVED** [review]

**Review by adam-a-a — APPROVED** [review]

**Pull request merged**